src/tools/snapshot/deserialization/DeserializationBuffer.java

```diff
@@ -1,91 +1,222 @@
 package tools.snapshot.deserialization;
 
+import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.nio.ByteOrder;
+import java.nio.channels.FileChannel;
+import java.util.ArrayList;
 
 import org.graalvm.collections.EconomicMap;
 
-import som.interpreter.actors.Actor;
+import som.interpreter.nodes.dispatch.CachedSlotWrite;
+import som.vm.VmSettings;
 import som.vmobjects.SClass;
+import som.vmobjects.SInvokable;
+import som.vmobjects.SObject;
 import tools.snapshot.SnapshotBackend;
 import tools.snapshot.deserialization.FixupInformation.FixupList;
+import tools.snapshot.nodes.ObjectSerializationNodes.SObjectSerializationNode.SlotFixup;
+import tools.snapshot.nodes.PrimitiveSerializationNodes.ClassSerializationNode;
 
 
 public class DeserializationBuffer {
 
-  private final ByteBuffer                wrapped;
+  protected final ByteBuffer              buffer;
   private final EconomicMap<Long, Object> deserialized;
   private long                            lastRef;
+  int                                     depth = 0;
+  private ArrayList<Long>                 unserialized;
 
   public DeserializationBuffer(final byte[] backing) {
-    wrapped = ByteBuffer.wrap(backing).asReadOnlyBuffer().order(ByteOrder.LITTLE_ENDIAN);
-    wrapped.rewind();
+    buffer = ByteBuffer.wrap(backing).asReadOnlyBuffer().order(ByteOrder.LITTLE_ENDIAN);
+    buffer.rewind();
     deserialized = EconomicMap.create();
+    unserialized = new ArrayList<>();
+  }
+
+  public DeserializationBuffer(final ByteBuffer buffer) {
+    this.buffer = buffer;
+    deserialized = EconomicMap.create();
+    unserialized = new ArrayList<>();
   }
 
   public byte get() {
-    return wrapped.get();
+    return buffer.get();
   }
 
   public void get(final byte[] b) {
-    wrapped.get(b);
+    buffer.get(b);
   }
 
   public short getShort() {
-    return wrapped.getShort();
+    return buffer.getShort();
   }
 
   public int getInt() {
-    return wrapped.getInt();
+    return buffer.getInt();
   }
 
   public long getLong() {
-    return wrapped.getLong();
+    return buffer.getLong();
   }
 
   public double getDouble() {
-    return wrapped.getDouble();
+    return buffer.getDouble();
+  }
+
+  public boolean allreadyDeserialized(final long reference) {
+    return deserialized.containsKey(reference);
+  }
+
+  public int getNumObjects() {
+    return deserialized.size();
+  }
+
+  private void printPosition(final long current) {
```

Comment:
This and the `printClass` methods look strange.
Are they used? Should they be implemented but have a VmSettings.DEBUG flag or something? I think there may be such a flag?

Reply:
I haven't used them for a while, we can probably remove them.
We could also use a flag, but this can produce a LOT of output.